clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1727,147 +1733,129 @@ class SyclKernelBodyCreator : public SyclKernelFieldHandler {
+    if (isArrayElement(FD, Ty))
+      return InitializedEntity::InitializeElement(SemaRef.getASTContext(),
+                                                  ArrayInfos.back().second,
+                                                  ArrayInfos.back().first);
+    return InitializedEntity::InitializeMember(FD, &VarEntity);
+  }
 
+  void addFieldInit(FieldDecl *FD, QualType Ty, MultiExprArg ParamRef) {
     InitializationKind InitKind =
         InitializationKind::CreateCopy(SourceLocation(), SourceLocation());
-    Expr *DRE = createInitExpr(FD);
-    InitializationSequence InitSeq(SemaRef, InitEntities.back(), InitKind, DRE);
-    ExprResult MemberInit =
-        InitSeq.Perform(SemaRef, InitEntities.back(), InitKind, DRE);
-    InitExprs.push_back(MemberInit.get());
-  }
-
-  void addArrayInit(FieldDecl *FD, int64_t Count) {
-    llvm::SmallVector<Expr *, 16> ArrayInitExprs;
-    for (int64_t I = 0; I < Count; I++) {
-      ArrayInitExprs.push_back(InitExprs.back());
-      InitExprs.pop_back();
-    }
-    std::reverse(ArrayInitExprs.begin(), ArrayInitExprs.end());
-    Expr *ILE = new (SemaRef.getASTContext())
-        InitListExpr(SemaRef.getASTContext(), SourceLocation(), ArrayInitExprs,
-                     SourceLocation());
-
-    // We need to find the type of the element for which we are generating the
-    // InitListExpr. For example, for a multi-dimensional array say a[2][3][2],
-    // the types for InitListExpr of the array and its 'sub-arrays' are -
-    // int [2][3][2], int [3][2] and int [2]. This loop is used to obtain this
-    // information from MemberExprBases. MemberExprBases holds
-    // ArraySubscriptExprs and the top of stack shows how far we have descended
-    // down the array. getDims() calculates this depth.
-    QualType ILEType = FD->getType();
-    for (int I = getDims(); I > 1; I--) {
-      const ConstantArrayType *CAT =
-          SemaRef.getASTContext().getAsConstantArrayType(ILEType);
-      assert(CAT && "Should only be called on constant-size array.");
-      ILEType = CAT->getElementType();
-    }
-    ILE->setType(ILEType);
-    InitExprs.push_back(ILE);
+    addFieldInit(FD, Ty, ParamRef, InitKind);
+  }
+
+  void addFieldInit(FieldDecl *FD, QualType Ty, MultiExprArg ParamRef,
+                    InitializationKind InitKind) {
+    InitializedEntity Entity = getFieldEntity(FD, Ty);
+
+    InitializationSequence InitSeq(SemaRef, Entity, InitKind, ParamRef);
+    ExprResult Init = InitSeq.Perform(SemaRef, Entity, InitKind, ParamRef);
+
+    InitListExpr *ParentILE = CollectionInitExprs.back();
+    ParentILE->updateInit(SemaRef.getASTContext(), ParentILE->getNumInits(),
+                          Init.get());
+  }
+
+  void addBaseInit(const CXXBaseSpecifier &BS, QualType Ty,
+                   InitializationKind InitKind) {
+    InitializedEntity Entity = InitializedEntity::InitializeBase(
+        SemaRef.Context, &BS, /*IsInheritedVirtualBase*/ false, &VarEntity);
```

Comment:
There WAS a TODO and I think it was right, it isn't STRICTLY correct.  That said, it seems to work?  I tried a while to come up with a case where it wouldn't in order to justify keeping the whole stack of entities (rather than just the arrays), but couldn't come up with one.

Reply:
I wonder if we'll need to once we stop decomposing classes. If kernel has a field with a base class I assume we can't initialize with VarEntity then right?

Reply:
Thats an interesting question... I don't see why we would need to in the non-decomp case, since those seem like a subset of this case.  But I am sure we'll see when we start the work on non-decomp.

IF thats the case, I suspect making ArrayInfos be simply CollectionInfos would be on the docket.

Reply:
>>> since those seem like a subset of this case
Maybe? Right now addBaseInit is called only when special types are used as a base class. I can't recall if we have tests which check if they can be used as base classes of fields of kernel as opposed to base class of functor itself. I understand the latter working but I'm not sure why the former works?

```
class A { }
class B : public A { }
class KernelFunctor {
       B obj;
}
```
VarEntity here is Entity for KernelFunctor object right? I'm not sure how using VarEntity when creating InitializedEntity for A works? If our special bases now works for similar code though, yay! I assume it would work for non-decomp case as well even if I don't understand why it works :)

Reply:
Right, it is.  I think these work, and I think they'll work for the non-decomp case (though, if I find a good example of it being necessary while implementing non-decomp, I'll obviously do it) for the same reason they work for the special types:  It seems that this entity isn't particularly important except in array cases.

I've got this in my notes as something to pay attention to in the near future as a part of the non-decomp patch/patches, so hopefully we'll know then.